### ISSUE TITLE
[Bug] Modifying note types while clock set incorrectly (on purpose, kids do this to parents phones sometimes...) results in a crash

### ISSUE THREAD

**Issue by Callum-0** [ankidroid/Anki-Android#8803]
###### Reproduction Steps
1. Modify a note type (Add/remove/rename field or delete note type)
2. Synchronise changes (If it doesn't crash)

###### Expected Result
- Modify note types without having the app crash.

###### Actual Result
- Modifying a note type results in a crash.

- When it doesn't end up crashing, it just reverses my changes when it finishes synchronising. Synchronising also no longer gives me the options to upload or download changes, instead it just downloads from the server and reverses my changes. 

###### Debug info
AnkiDroid Version = 2.14.6 
Android Version = 11 
ACRA UUID = 758dae2e-58aa-40d0-9d37-18809c5cb0dc 
Scheduler = std 
Crash Reports Enabled = true

###### Research
- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [x] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

###### Additional Information
`java.lang.RuntimeException
at com.ichi2.anki.ModelBrowser$DeleteModelHandler.actualOnPostExecute(ModelBrowser.java:5)
at com.ichi2.anki.ModelBrowser$DeleteModelHandler.actualOnPostExecute(ModelBrowser.java:1)
at com.ichi2.async.TaskListenerWithContext.onPostExecute(TaskListenerWithContext.java:2)
at com.ichi2.async.CollectionTask.onPostExecute(CollectionTask.java:4)
at com.ichi2.async.CollectionTask.onPostExecute(CollectionTask.java:1)
at android.os.AsyncTask.finish(AsyncTask.java:771)
at android.os.AsyncTask.access$900(AsyncTask.java:199)
at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:788)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:246)
at android.app.ActivityThread.main(ActivityThread.java:8443)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:596)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)`

`java.lang.RuntimeException: An error occurred while executing doInBackground()
at android.os.AsyncTask$4.done(AsyncTask.java:415)
at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
at java.util.concurrent.FutureTask.run(FutureTask.java:271)
at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:305)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:923)
Caused by: java.lang.AssertionError: Mod was assumed to be already changed, but is not
at com.ichi2.utils.Assert.that(Assert.java:3)
at com.ichi2.libanki.Models.addFieldModChanged(Models.java:1)
at com.ichi2.async.CollectionTask.doInBackGroundAddField(CollectionTask.java:6)
at com.ichi2.async.CollectionTask.actualDoInBackground(CollectionTask.java:22)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:2)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:1)
at android.os.AsyncTask$3.call(AsyncTask.java:394)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)`

**Comment by Callum-0** [comment]
Hi @mikehardy ! Thanks for the response. I just tested it on the parallel A build, but the issue is still happening. It won't allow me to synchronise on the parallel build either, it results in a crash.

**Comment by mikehardy** [comment]
:thinking: 

I don't see the crash reports for the parallel build, but I see these for the old ones.

Thanks for trying the new build to see how it went.

There *may* be resolution if you run check database on the desktop side (*assuming it is okay to full sync from desktop*) then full sync to ankiweb and then to ankidroid to get the results of a checked database. Not sure if that is an option for you, and it's vague either way - we've just never seen this reported and that may fix it.

Either way it's logged now with stack traces so we should be able to figure it out

```
java.lang.RuntimeException
at com.ichi2.anki.ModelBrowser$DeleteModelHandler.actualOnPostExecute(ModelBrowser.java:5)
at com.ichi2.anki.ModelBrowser$DeleteModelHandler.actualOnPostExecute(ModelBrowser.java:1)
at com.ichi2.async.TaskListenerWithContext.onPostExecute(TaskListenerWithContext.java:2)
at com.ichi2.async.CollectionTask.onPostExecute(CollectionTask.java:4)
at com.ichi2.async.CollectionTask.onPostExecute(CollectionTask.java:1)
at android.os.AsyncTask.finish(AsyncTask.java:771)
at android.os.AsyncTask.access$900(AsyncTask.java:199)
at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:788)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:246)
at android.app.ActivityThread.main(ActivityThread.java:8443)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:596)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
at android.os.AsyncTask$4.done(AsyncTask.java:415)
at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
at java.util.concurrent.FutureTask.run(FutureTask.java:271)
at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:305)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:923)
Caused by: java.lang.AssertionError: Mod was assumed to be already changed, but is not
at com.ichi2.utils.Assert.that(Assert.java:3)
at com.ichi2.libanki.Models.addFieldModChanged(Models.java:1)
at com.ichi2.async.CollectionTask.doInBackGroundAddField(CollectionTask.java:6)
at com.ichi2.async.CollectionTask.actualDoInBackground(CollectionTask.java:22)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:2)
at com.ichi2.async.CollectionTask.doInBackground(CollectionTask.java:1)
at android.os.AsyncTask$3.call(AsyncTask.java:394)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
... 4 more
```

**Comment by mikehardy** [comment]
That new UUID does have a crash, this time it is in a different area, so the new code is still crashing for some reason but not at all the same as the old code:

```
java.lang.ClassCastException: java.lang.Object[] cannot be cast to java.lang.Long
at com.ichi2.anki.DeckPicker$3.onPostExecute(DeckPicker.java:54)
at com.ichi2.async.Connection.onPostExecute(Connection.java:6)
at com.ichi2.async.Connection.onPostExecute(Connection.java:1)
at android.os.AsyncTask.finish(AsyncTask.java:771)
at android.os.AsyncTask.access$900(AsyncTask.java:199)
at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:788)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loop(Looper.java:246)
at android.app.ActivityThread.main(ActivityThread.java:8443)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:596)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

**Comment by david-allison-1** [comment]
Before you start again, could you provide us with the file located at `AnkiDroid/collection.anki2` file so we can diagnose this issue.

This is the Anki database, which contains everything except media

If your collection didn't contain sensitive information, could you rename it to a .zip and upload it to GitHub

If it does contain sensitive information, could you send it to my Gmail: `davidallisongithub`

Thanks

David

**Comment by Callum-0** [comment]
@mikehardy - I have seemed to fix the bug. I just tried to reinstall the alpha build and this time, it gave me an error message saying that my clock was ~350 seconds behind. So I corrected the clock and now it synchronises everything correctly!

It never gave me that error message before when I had tried to reinstall AnkiDroid, even on older versions. I can confirm that the clock was the issue (see the video below), because I just put the clock back a couple minutes again, and after that it starts crashing again. But this time, it doesn't display that error message letting me know that my clock is incorrect, so I think that should be fixed.

Sorry for the inconvenience, and thanks for the help! 


https://user-images.githubusercontent.com/84001785/118018640-bb90e680-b360-11eb-96d5-6ab82dcd70af.mp4

**Comment by mikehardy** [comment]
That is so strange! There is no reason we should crash based on a clock setting. It's actually a classic for children to take their parent's phone and hack their way around restrictions by monkeying with the time, apps can't expect reliable time from the device. Thanks for noting that.

Glad you're moving again!

**Comment by dae** [comment]
Assuming this is fixed; if you can reproduce it in 2.17 (alpha+ when it's released), please let us know.
